lib/kdl/include/kdl/vector_utils.h

```diff
@@ -0,0 +1,381 @@
+     * @param s2 the second set
+     * @return a vector containing the set union of s1 and s2.
+     */
+    template<typename S1, typename S2>
+    auto set_union(const S1& s1, const S2& s2) {
+        using T1 = typename S1::value_type;
+        using T2 = typename S2::value_type;
+        using C1 = typename S1::value_compare;
+        using C2 = typename S2::value_compare;
+        static_assert(std::is_same<C1, C2> ::value, "incompatible comparators");
+
+        using T = typename std::common_type<T1, T2>::type;
+        using C = C1;
+
+        std::vector<T> result;
+        result.reserve(s1.size() + s2.size());
+        std::set_union(std::begin(s1), std::end(s1), std::begin(s2), std::end(s2), std::back_inserter(result), C());
+        return result;
+    }
+
+
+    /**
+     * Returns a vector containing all values from s1 and s2 which are present in both sets.
+     *
+     * Expects that both S1 and S2 declare the types of their values with ::value_type and the comparator used to
+     * compare the values with ::value_compare. Additionally, the type of value_compare must be identical in both sets.
+     *
+     * The value type of the returned vector is the common type of S1 and S2's member types. The order of the values in
+     * the returned vector complies with the common comparator of s1 and s2.
+     *
+     * @tparam S1 the type of the first set
+     * @tparam S2 the type of the second set
+     * @param s1 the first set
+     * @param s2 the second set
+     * @return a vector containing the set union of s1 and s2.
+     */
+    template<typename S1, typename S2>
+    auto set_intersection(const S1& s1, const S2& s2) {
+        using T1 = typename S1::value_type;
+        using T2 = typename S2::value_type;
+        using C1 = typename S1::value_compare;
+        using C2 = typename S2::value_compare;
+        static_assert(std::is_same<C1, C2> ::value, "incompatible comparators");
+
+        using T = typename std::common_type<T1, T2>::type;
+        using C = C1;
+
+        std::vector<T> result;
+        result.reserve(s1.size() + s2.size());
+        std::set_intersection(std::begin(s1), std::end(s1), std::begin(s2), std::end(s2),
+            std::back_inserter(result), C());
+        return result;
+    }
+
+    /**
+     * Clears the given vector and ensures that it has a capacity of 0 afterwards.
+     *
+     * @tparam T the type of the vector elements
+     * @param v the vector
+     */
+    template<typename T>
+    void clear_to_zero(std::vector<T>& v) {
+        v.clear();
+        v.shrink_to_fit();
+    }
+
+    /**
+     * Applies the given deleter to every element of the given vector, and clears the vector afterwards.
+     *
+     * @tparam T the type of the vector elements
+     * @param v the vector
+     */
+    template<typename T, typename D = deleter<T*>>
+    void clear_and_delete(std::vector<T*>& v, const D& deleter = D()) {
+        kdl::delete_all(v, deleter);
+        v.clear();
+    }
+}
```

Comment:
I'd suggest adding some signed-int accessors:
`T at(const std::vector<T>& v, int index);`
`T at(const std::vector<T>& v, int64_t index);`

also generic collection size getters:
`int size_int(const C& collection);`
`int64_t size_int64(const C& collection);`

Whenever we have static_casts from signed ints <-> size_t, this is a code smell, and there's a lot of these in TB!
(this should be a separate issue I guess.)

Reply:
I'll add these methods, but I think most of the type conversions are because Qt uses int for indexing / size.

Reply:
Which I believe is the actual code smell here ;-)

Reply:
done

Reply:
cool, thanks. 

I blame the stl using unsigned, though. The static casts to work around this crop up in TB not just in code interacting with Qt.

there's a video that's worth checking out where this topic comes up a few times, that is the best discussion of this I've seen, links in this post 
https://github.com/ericniebler/stl2/issues/182#issuecomment-287683189

Reply:
It looks like everyone is disagreeing with me ;-)